kits/arm/ex_impedance_control_damping.py

```diff
@@ -0,0 +1,223 @@
+mode = -1
+prevmode = -1
+
+# 0: Overdamped
+overdamped_kp = np.array([100, 100, 0, 5, 5, 1])
+overdamped_kd = np.array([15, 15, 0, 0, 0, 0])
+
+# 1: Critically damped
+critically_damped_kp = np.array([100, 100, 0, 5, 5, 1])
+critically_damped_kd = np.array([5, 5, 0, 0, 0, 0])
+
+# 2: Underdamped
+underdamped_kp = np.array([100, 100, 0, 5, 5, 1])
+underdamped_kd = np.array([0, 0, 0, 0, 0, 0])
+
+damping_kp = [overdamped_kp, critically_damped_kp, underdamped_kp]
+damping_kd = [overdamped_kd, critically_damped_kd, underdamped_kd]
+
+# Increase feedback frequency since we're calculating velocities at the
+# high level for damping. Going faster can help reduce a little bit of
+# jitter for fast motions, but going slower (100 Hz) also works just fine
+# for most applications.
+arm.group.feedback_frequency = 200.0
+
+# Double the effort gains from their default values, to make the arm more sensitive for tracking force.
+# TODO
+
+enable_logging = True
+goal = hebi.arm.Goal(arm.size)
+
+# Start background logging
+if enable_logging:
+    arm.group.start_log('dir', 'logs', mkdirs=True)
+
+print('Commanded gravity-compensated zero force to the arm.')
+print('  💪 (B2) - Toggles an impedance controller on/off:')
+print('            ON  - Apply controller based on current position')
+print('            OFF - Go back to gravity-compensated mode')
+print('  📈 (B1) - Exits the demo, and plots graphs. May take a while.')
+
+controller_on = False
+
+# while button 1 is not pressed
+while not m.get_button_state(1):
+
+    # if not arm.update():
+    #     print("Failed to update arm")
+    #     continue
+    arm.update()
+
+    arm.send()
+
+    if m.update(timeout_ms=0):
+
+        # Set and unset impedance mode when button is pressed and released, respectively
+        if (m.get_button_diff(2) == 1):
+
+            controller_on = True
+            arm.set_goal(goal.clear().add_waypoint(position=arm.last_feedback.position))
+
+        elif (m.get_button_diff(2) == -1):
+            
+            controller_on = False
+
+        # If in impedance mode set led blue
+        m.set_led_color("blue" if controller_on else "green", blocking=False)
+
+    if not controller_on:
+        arm.cancel_goal()
+        mode = -1 # Free
+        prevmode = -1
+
+    # Check when end-effector travels below the floor, for floor demo
+    else:
+
+        # Use forward kinematics to calculate pose of end-effector
+        ee_pose_curr = np.eye(4)
+        arm.robot_model.get_end_effector(arm.last_feedback.position, ee_pose_curr)
+
+        # Assign mode based on current position
+        for i in range(len(lower_limits)):
+
+            if ee_pose_curr[2,3] > lower_limits[i]:
+
+                mode = i
+        
+        # Change gains only upon mode switches
+        if not mode == prevmode and mode >= 0:
+
+            impedance_controller.set_kp(damping_kp[mode][0], damping_kp[mode][1], damping_kp[mode][2], damping_kp[mode][3], damping_kp[mode][4], damping_kp[mode][5])
```

Comment:
I would use the spread operator here to make things less verbose:
```
impedance_controller.set_kp(*damping_kp[mode])
impedance_controller.set_kd(*damping_kd[mode])
```

Reply:
All done!